File: scripts/create-google-ads-api-design-pdf.mjs
import { mkdirSync, writeFileSync } from "node:fs";

const out = "output/pdf/tippos-google-ads-api-design.pdf";
const lines = [
  "tippos Market Intelligence",
  "Google Ads API Basic Access - Internal Tool Design",
  "",
  "Purpose",
  "tippos is building an internal market-intelligence workflow for its own product",
  "and SEO decision-making. It researches aggregate Google Search demand for",
  "tip-calculator and tipping-related topics in US English.",
  "",
  "Intended users and access",
  "Only authorized tippos employees and contractors use this internal workspace.",
  "External users, clients, advertisers, and the public do not receive API access,",
  "credentials, or raw Google Ads API results.",
  "",
  "API use",
  "The planned integration uses Google Ads keyword-planning services to generate",
  "candidate phrases and request historical metrics only for an internally reviewed",
  "list. Requests target the United States, English, and Google Search.",
  "",
  "Excluded activity",
  "The workflow does not create or optimize campaigns, budgets, bids, ads, audiences,",
  "conversion tracking, or remarketing. It does not manage advertising for third parties.",
  "",
  "Data handling and safeguards",
  "The integration records aggregate keyword metrics only. It is not designed to collect",
  "personal data, user-level search histories, or Google Ads performance data. API",
  "credentials are server-side secrets and are not exposed in the browser interface.",
  "Keyword ideas are candidates, not evidence, until Google Ads returns historical metrics.",
  "",
  "System design",
  "1. An authorized internal user starts a server-side research run.",
  "2. The server authenticates with a service account and tippos developer token.",
  "3. Discovery creates candidates only; measurement requests approved historical metrics.",
  "4. Returned results are validated and stored in the separate Market Intelligence",
  "   Supabase project for internal product and SEO review.",
  "",
  "Operational controls",
  "Production use begins only after approved access and keyword-research permission.",
  "Requests are rate limited, results are refreshed monthly, and provider limitations",
  "and collection time are retained for auditability.",
  "",
  "Current rollout state (2026-07-26)",
  "The design is prepared, but the credential connection is not deployed and no Google",
  "Ads API keyword metrics have been imported into the market-intelligence database.",
];

const esc = (s) => s.replace(/\\/g, "\\\\").replace(/\(/g, "\\(").replace(/\)/g, "\\)");
let content = "BT\n/F1 10 Tf\n50 750 Td\n";
lines.forEach((line, i) => {
  const font = i === 0 ? 18 : (line && ![4, 9, 14, 19, 23, 30, 36, 41].includes(i) ? 12 : 10);
  if (i === 0) content += "/F1 18 Tf\n";
  else if ([3, 8, 13, 18, 22, 29, 35, 40].includes(i)) content += "/F1 12 Tf\n";
  else content += "/F1 10 Tf\n";
  content += `(${esc(line)}) Tj\n0 -${i === 0 ? 28 : 14} Td\n`;
});
content += "ET";
const objects = [
  "<< /Type /Catalog /Pages 2 0 R >>",
  "<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
  "<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>",
  "<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
  `<< /Length ${Buffer.byteLength(content, "ascii")} >>\nstream\n${content}\nendstream`,
];
let pdf = "%PDF-1.4\n";
const offsets = [0];
objects.forEach((obj, index) => { offsets.push(Buffer.byteLength(pdf, "ascii")); pdf += `${index + 1} 0 obj\n${obj}\nendobj\n`; });
const start = Buffer.byteLength(pdf, "ascii");
pdf += `xref\n0 ${objects.length + 1}\n0000000000 65535 f \n`;
offsets.slice(1).forEach((offset) => { pdf += `${String(offset).padStart(10, "0")} 00000 n \n`; });
pdf += `trailer\n<< /Size ${objects.length + 1} /Root 1 0 R /Title (tippos Google Ads API Design) >>\nstartxref\n${start}\n%%EOF\n`;
mkdirSync("output/pdf", { recursive: true });
writeFileSync(out, pdf, "ascii");
console.log(out);

File: scripts/create-google-ads-api-design-pdf.py
from pathlib import Path

from reportlab.lib.colors import HexColor
from reportlab.lib.pagesizes import letter
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import inch
from reportlab.platypus import KeepTogether, Paragraph, SimpleDocTemplate, Spacer


OUTPUT = Path("output/pdf/tippos-google-ads-api-design.pdf")
OUTPUT.parent.mkdir(parents=True, exist_ok=True)

styles = getSampleStyleSheet()
styles.add(ParagraphStyle(
    name="TitleTippos", parent=styles["Title"], fontName="Helvetica-Bold",
    fontSize=24, leading=29, textColor=HexColor("#111111"), spaceAfter=10,
))
styles.add(ParagraphStyle(
    name="SubTitleTippos", parent=styles["Normal"], fontName="Helvetica",
    fontSize=10, leading=14, textColor=HexColor("#5B5B5B"), spaceAfter=22,
))
styles.add(ParagraphStyle(
    name="HeadingTippos", parent=styles["Heading2"], fontName="Helvetica-Bold",
    fontSize=13, leading=17, textColor=HexColor("#242424"), spaceBefore=12, spaceAfter=5,
))
styles.add(ParagraphStyle(
    name="BodyTippos", parent=styles["BodyText"], fontName="Helvetica",
    fontSize=10, leading=14, textColor=HexColor("#2D2D2D"), spaceAfter=7,
))


def paragraph(text, style="BodyTippos"):
    return Paragraph(text, styles[style])


sections = [
    ("Purpose", "tippos is building an internal market-intelligence workflow for its own product and SEO decision-making. The workflow researches aggregate Google Search demand for tip-calculator and tipping-related topics in US English. It is not a public keyword tool, a customer-facing product, or an advertising-management service for other businesses."),
    ("Intended users and access", "Only authorized tippos employees and contractors who support product, SEO, or market research will use the internal workspace. External users, clients, advertisers, and the general public will not receive access to the API tool, developer token, credentials, or raw Google Ads API results."),
    ("API use", "The planned integration uses Google Ads API keyword-planning services for internal research. It generates candidate phrases from a small topic-specific seed set, labels those ideas as unmeasured candidates, and requests historical metrics only for an internally reviewed list. Requests are targeted to the United States, English, and Google Search. Returned aggregate metrics retain their month, geography, language, network, canonical keyword, and close-variant metadata."),
    ("Excluded activity", "The workflow does not create, modify, or optimize Google Ads campaigns, budgets, bids, ads, audiences, conversion tracking, or remarketing. It does not automate advertising for any third party."),
    ("Data handling and safeguards", "The integration records only aggregate keyword metrics returned by Google Ads API. It is not designed to collect personal data, user-level search histories, or Google Ads account performance data. API credentials are held as server-side secrets and are not exposed in the internal browser interface. Keyword ideas remain candidates rather than evidence until Google Ads returns historical metrics. Empty provider responses do not activate a source or become a success claim."),
    ("System design", "1. An authorized internal user starts a server-side research run.<br/>2. The server authenticates to Google Ads API with a service account and the tippos developer token.<br/>3. Discovery may retrieve candidate phrases but does not write measured metrics.<br/>4. Measurement requests historical metrics for an approved bounded list.<br/>5. Returned results are validated, deduplicated, and stored in the separate tippos Market Intelligence Supabase project.<br/>6. Internal users review aggregate research outputs to inform product and SEO decisions."),
    ("Operational controls", "Production use begins only after the developer token receives the required approved access and permissible use for keyword research. Requests are rate limited and retried conservatively. Research is refreshed monthly rather than used as real-time reporting. The implementation keeps an audit trail of source, collection time, and provider limitations."),
    ("Current rollout state", "The integration design and validation are prepared, but the Google Ads credential connection is not deployed. No Google Ads API keyword metrics have been imported into the market-intelligence database as of 2026-07-26."),
]

story = [
    paragraph("tippos Market Intelligence", "TitleTippos"),
    paragraph("Google Ads API Basic Access - Internal Tool Design", "SubTitleTippos"),
]
for heading, body in sections:
    story.append(KeepTogether([paragraph(heading, "HeadingTippos"), paragraph(body)]))

doc = SimpleDocTemplate(
    str(OUTPUT), pagesize=letter, rightMargin=0.72 * inch, leftMargin=0.72 * inch,
    topMargin=0.68 * inch, bottomMargin=0.68 * inch,
    title="tippos Google Ads API Design",
    author="tippos",
)
doc.build(story)
print(OUTPUT)
